### ISSUE TITLE
Output from batched pipeline is out of order

### ISSUE THREAD

**Issue by Narasimha1997** [NVIDIA/DALI#3692]
Hi, I am trying a usecase where I am using DALI for some pre-processing on a batch of 16 images before feeding it as input to an object detection model. The usecase I am working on demands batch inputs to be in the order throughout the pipeline. For example, If image 4 is in the batch entry at index 3, then I want the image to be at the same entry even after the pipeline so I can match the detection bounding boxes with the original image later. 

For this I wrote a simple preprocessing pipeline:
```python
@pipeline_def
def image_decoder(ext_s=None):
    jpeg_frames = fn.external_source(source=ext_s, batch=True)

    decoded = fn.decoders.image(jpeg_frames, device='mixed')
    normalized = fn.normalize(decoded, stddev=255., device='gpu')
    transposed = fn.transpose(normalized, perm=(2, 0, 1), device='gpu')
    return transposed
```
Created a simple iterator that holds the array of 16 JPEG bytes (uint8 numpy array)
```python
class FrameHolder:

    def __init__(self):
        self.frames = [None]
    
    def __iter__(self):
        self.i = 0
        return self
    
    def get_frames(self):
        yield self.frames
    
    def __next__(self):
        return self.frames
```
Then I init. both pipeline and iterator as follows:

```python
self.iterator = FrameHolder()
self.pipeline = image_decoder(
        batch_size=self.batch_size,
        ext_s=self.iterator,
        num_threads=1,
        device_id=0
)
self.pipeline.build()
```

And during inference, I am running the pipeline like this:
```python
self.iterator.frames = frames
batches = self.pipeline.run()
batches = batches[0].as_tensor()
torch_tensor = torch.zeros((self.batch_size, 3, width, height)).cuda()

dali_torch.feed_ndarray(batches, torch_tensor)
predictions = self.model(torch_tensor.half(), False)[0]
```

But I notice that the prediction at batch index doesn't match the position of the original image in the batch. That means, the outputs of the pipeline `batches` is shuffled somewhere in the pipeline in random order. I tried with a traditional cv2 + numpy pipeline for image processing and it works fine (so the model is not shuffling the order of predictions in any way), which makes it impossible to keep track of the original positions of the images in the array before feeding it to the batch. I looked at the documentation and other possible sources, but not able to find solution for this. Is there any way we can ensure ordering of elements in the batch when it flows through the pipeline? If yes, can we document it somewhere so it can help others. If no, is there any alternative way to do this?

**Comment by mzient** [comment]
Hello @Narasimha1997 , thanks for reaching out to DALI team.
The problem you're encountering is not that the batch is shuffled; rather, it is a problem of prefetching in DALI. When you create a DALI pipeline with default configuration, it will consume one batch ahead (on first run) so when you call `run` it can immediately start processing the next batch - that's essentially double buffering. What you try to do is to swap in the batch just prior to a call to `run`, but that batch will be used for the _next_ run. The data you see is not shuffled - it is delayed by one batch.
If your use case requires you to prepare the data immediately before calling run (and you cannot just craft it inside your `source`), then your only option is to limit the `prefetch_queue_depth` to 1. See the following example:

You can run this example:
```Python
import nvidia.dali as dali
import nvidia.dali.fn as fn
import numpy as np

class my_iter:
    def __init__(self):
        self.seq = 0
        self.data = []
        
    def __iter__(self):
        print("Resetting")
        self.seq = 0
        return self
        
    def __next__(self):
        print(f"Getting batch #{self.seq}")
        self.seq += 1
        return self.data
    

batch_size = 4
prefetch_depth = 2  # that's the default - try putting 1 here and see that data is in sync

@dali.pipeline_def(batch_size=batch_size, device_id=0, num_threads=3, prefetch_queue_depth=prefetch_depth)
def my_pipe(iter):
    return fn.external_source(iter, batch=True)

def format_tensor_list(tensor_list):
    out = "["
    for t in tensor_list:
        out += f"{np.array(t)}, "
    out += "]"
    return out
    
it = my_iter()
start = 0
it.data = [np.float32([i]) for i in range(start, start+batch_size)]
print("Setting data:     ", format_tensor_list(it.data))
pipe=my_pipe(it)
print("Build")
pipe.build()
print("Run")
o, = pipe.run()
print("Output:           ", format_tensor_list(o))

start += batch_size
new_data = [np.float32([i]) for i in range(start, start+batch_size)]
print("\n")
print("Setting new data: ", format_tensor_list(new_data))

it.data = new_data

print("Run")
o, = pipe.run()
print("Output:           ", format_tensor_list(o))


start += batch_size
new_data = [np.float32([i]) for i in range(start, start+batch_size)]
print("\n")
print("Setting new data: ", format_tensor_list(new_data))

it.data = new_data

print("Run")
o, = pipe.run()
print("Output:           ", format_tensor_list(o))
```

This will cripple performance, however. If there's some label data associated with your pipeline input, you can just pass your bounding box data (or an index to identify it by) to the pipeline and pass it to an auxiliary output.
```Python
# my_data_source must now produce a batch of frames and indices
encoded index = fn.external_source(source=my_data_source, num_outputs=2) 
...  # do all the stuff here
return transposed, index
```
